File: packages/javascript/lib/web/escapeHtml.ts
// Compiled once. Building a `RegExp` inside a function recompiles the pattern on every call.
const UNESCAPED_HTML_CHARACTERS = /[&<>"']/g;

// `'` is written as `&#39;` rather than `&apos;`, which HTML 4 did not define and which
// therefore does not survive every parser.
const HTML_ESCAPES: { [key: string]: string } = {
	'&': '&amp;',
	'<': '&lt;',
	'>': '&gt;',
	'"': '&quot;',
	"'": '&#39;'
};

export function escapeHtml(str: string): string {
	if (!str) {
		return '';
	}

	return str.replace(UNESCAPED_HTML_CHARACTERS, (char: string) => HTML_ESCAPES[char]);
}

File: packages/python/qsu/web/escapeHtml.py
import re

# Compiled once. Building a pattern inside a function recompiles it on every call.
_unescapedHtmlCharacters = re.compile(r'''[&<>"']''')

# `'` is written as `&#39;` rather than `&apos;`, which HTML 4 did not define and which
# therefore does not survive every parser. The built-in `html.escape` writes `&#x27;`, so
# this is not a wrapper around it.
_htmlEscapes = {
	'&': '&amp;',
	'<': '&lt;',
	'>': '&gt;',
	'"': '&quot;',
	"'": '&#39;',
}


def escapeHtml(str: str) -> str:
	if not str:
		return ''

	return _unescapedHtmlCharacters.sub(lambda m: _htmlEscapes[m.group(0)], str)
